accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+type).
+
+```plain
+
+    ;; `Iadd` is a term symbol, and `(Iadd ...)` terms take two arguments
+    ;; of type `Value` and produce a result of type `Value`.
+    (decl (Iadd Value Value) Value)
+    
+    (decl (Iconst u32) Value)
+```
+
+Now, given a proper type environment for values and terms, we need to
+define ways to process them: with extractors and constructors, and
+with internal rules. A particular term can *either* have an external
+extractor and/or constructor function, or can have an internal rule,
+but never both. If a term is to be defined by an extractor and/or
+constructor, it must be declared with an `extern` like so:
+
+```plain
+    (decl extern (MyTerm u32) u32)
+```
+
+## Defining External Extractors and Constructors
+
+First, we define *extractors* and *constructors* for some terms. These
+terms are those that represent the external interface to the
+instruction selector rules. Extractors, as introduced above, provide a
+view onto a virtual term tree at the input.
+
+One can think of an extractor like a "reverse function": it is given
+the value of the whole term tree that the extractor is meant to match,
+and if it matches successfully, it returns values for each of the
+arguments. Extractors are attached to terms as follows:
+
+```plain
+
+    (decl extern (concat u32 u32) u64)
+    (extractor concat "unconcat")
+```
+
+This indicates to the DSL compiler that the `concat` term can be used
+in the left-hand side of a rule (as described below), and when a match
+is evaluated, a Rust function `unconcat` will be called. This external
+function is assumed to have the signature:
+
+```rust
+
+    fn unconcat<C>(context: &mut C, value: u64) -> Option<(u32, u32)>;
+```
+
+If the function returns a `Some((a, b))` value, the match was a
+success and the arguments' values are recursively matched by
+sub-patterns.
+
+Similarly, a constructor can be attached to a term as follows:
+
+```plain
+
+    (decl extern (concat u32 u32) u64)
+    (constructor concat "concat")
+```
+
+which implies the existence of a Rust function:
+
+```rust
+
+    fn concat<C>(context: &mut C, val1: u32, val2: u32) -> u64;
+```
+
+Note that a constructor, unlike an extractor, cannot fail. More on the
+execution model below.
+
+## Defining Rules
+
+The final way to define a term -- and the most common -- is to provide
+a set of pattern-matching rules for it. A rule consists of two parts,
+the *left-hand side* and the *right-hand side*. The rule specifies
+that whenever a term of form that matches the left-hand side exists
+and is to be reduced, it can be rewritten as the right-hand side.
+
+Several examples of rule definitions follow:
+
+```plain
+
+    (rule (A x y z) (B x y z))  ;; rewrite (A x y z) to (B x y z)
+    
+    (rule (A x @ (B 42 _))  ;; match (A (B 42 _)), binding x to the B subterm;
+                            ;; the `42` matches a constant value;
+                            ;; the `_` matches anything (wildcard).
```

Comment:
The wildcard only exists in the grammar for LHS -- is it expected that it would always be discarded entirely (and if so, do you have an example of actual rules that would unconditionally discard one input), or is there a way to bind it and reuse it in the RHS that i've missed?

Reply:
ah wait, in this example the wildcarded input is actually bound to `x`, so nevermind me.

Reply:
> do you have an example of actual rules that would unconditionally discard one input

It could be used in something like "anything times zero is zero":

```lisp
(rule (Imul _ 0)
      (Iconst 0))
```